### ISSUE TITLE
Show the data in a table on W&B

### ISSUE THREAD

**Issue by jxmorris12** [jscuds/rf-bert#17]
I want to see some data. When we do the logging, we should also log some sample points from both train and val, showing their actual text and label, as well as the associated loss. We should do this at least for retrofitting, might not be as important for fine-tuning.

**Comment by jscuds** [comment]
Sounds good.  Ultimately I'll use the `wandb.Table()` call and have the following columns:
- split
- text (see Q1 below)
- label(s) - see Q2 below
- loss - see Q3 below

Questions:
1. I wrote a hacky "decoder" for elmo that takes a sentence tensor and returns a string representation.  I attached it (as a text file) [decode_utils.txt](https://github.com/jscuds/rf-bert/files/8102929/decode_utils.txt) --> should I include it in utils.py?
2. I'm not sure I know what the label for a retrofitting point is.  
    a. The predictions are the embedded word representations of the two *positive* words and the two *negative* words; so the label is what they _should_ be...which I don't think we know, correct?
    b. I could show the results of the `model.forward()` call: the shared word/index...but I don't think that's what you're looking for.
3. What's the easiest way to pull the loss for a single prediction:
    a. Add another list in the `RetrofitExperiment.retrofit_hinge_loss()` method that just grabs a random example and its loss before collapsing it along the mean of all examples in a batch?
    b. Re-passing a single example through the model with `torch.no_grad()` and getting the loss for that one point?
4. What's a good number of points?  I figure start with 10 and we can pass it as an `argparse` argument like `num_table_examples`.

**Comment by jxmorris12** [comment]
1. yeah, utils is a good place for this.
2. Good point, I just meant positive or negative, i.e. you should log them in separate rows, with a column that says positive or negative, something like: [sentence1, sentence2, distance, positive/negative]
3. Ah this is kind of a hard question. Maybe you could change the hinge loss to return the totals (instead of the means) and that would make your life easier? Recomputing inside a torch.no_grad() isn't the worst idea, if this only happens at most a few times per epoch, but would prefer the other way.
4. Yeah 10 sounds good, definitely want it in argparse!

Btw, the reason I posted this issue is because you pointed out that the table in the paper doesn't really correspond to the average distances we're seeing, so hopefully by logging some individual datapoints we'll get an idea of how that table was created. (also it's always nice to look at the data like this)

**Comment by jscuds** [comment]
Current plan - let me know (at your leisure) if it doesn't nest with your concept.
1. The whole goal would be to track the same example set through each epoch.  In order to guarantee tracking across epochs I can grab the `num_table_examples` positive pairs at random from the dataset class before training.  The negative pairs will be different each epoch (the "negative" sentence is selected at random in the `__getitem__` method of the dataset class).
2. Each example would have two rows per epoch:
    a. _Row 1:_ [sent1, sent2, target_word, positive, pos_distance, hinge_loss]
    b. _Row 2:_  [sent1, sent2, target_word, negative, neg_distance, hinge_loss] (
    c. sent1 or sent2 will be the same as Row 1, and the other will be the negative sampled sentence
    d. hinge losses will be the same in both rows
3. At the end of training, we should have two tables:
    a. **"Train Samples"** w/ length = `num_table_examples` * 2 * `epochs`
    b. **"Validation Samples"** w/ length = `num_table_examples` * 2 * `epochs`

Another note (re: #3 in comment above) - at this point it might be easier to add an optional argument to `retrofit_hinge_loss` that just passes the index of the example we care about, and track that individual loss.  I'll figure it out.

**Comment by jscuds** [comment]
Closed w/ #18
